### PR TITLE
SRC-4962 Switch controllers in change_control_type_callback

### PR DESCRIPTION
**Pull request by mykolasjuraitis**
## Proposed changes

Modify change_control_type_callback in the following way:
- Callback will no longer block real time control loop by sharing mutex with it
- Callback will no longer block real time control loop by requesting to switch controllers
- Callback will span a background thread to switch controllers (can't call other services from inside callback)
- Before controllers are switched "nullify demand" mode will be activated which will send 0 to the motors during switching
- Controllers will be switched one by one instead of all together to minimise strain on computer resources
- If controller being started is not in loaded controller list it will be loaded
- If all controllers are switched successfully "nullify demand" mode will be reset to the previous value
- if any of controllers failed to switch, then hand will stay in "nullify demand" mode

Python code calling this service is changed not to do controller switching in a separate PR as it is in a different repository (see https://github.com/shadow-robot/sr_interface/pull/647).

2 lint errors were ignored:
- `#include <thread>` causes: <thread> is an unapproved C++11 header.  [build/c++11]
- Creating a thread with multi-line expression causes one of these 2:
  - } should be on a line by itself  [whitespace/braces]
  - Closing ) should be moved to the previous line  [whitespace/parens] 

## Checklist

If the task is applicable to this pull request (see applicability criteria in brackets), make sure it is completed before checking the corresponding box. Otherwise, tick the box right away. Make sure that **ALL** boxes are checked **BEFORE** the PR is merged.

- [x] Manually tested that added code works as intended (if any functional/runnable code is added).
- [x] Added automated tests (if a new class is added (Python or C++), interface of that class must be unit tested).
- [x] Tested on real hardware (if the changed or added code is used with the real hardware).
- [x] Added documentation (For any new feature, explain what it does and how to use it. Write the documentation in a relevant space, e.g. Github, Confluence, etc.)

### CODE REVIEW

File: sr_robot_lib/src/sr_motor_robot_lib.cpp [hunk 1]
```diff
@@ -1197,100 +1194,110 @@ namespace shadow_robot
...
+      switch_controllers(request.control_type);
     }
-
     response.result = control_type_;
     return true;
   }
 
   template<class StatusType, class CommandType>
-  bool SrMotorRobotLib<StatusType, CommandType>::change_control_parameters(int16_t control_type)
+  void SrMotorRobotLib<StatusType, CommandType>::switch_controllers(sr_robot_msgs::ControlType control_type)
```
Comment (carebare47) [inline sr_robot_lib/src/sr_motor_robot_lib.cpp:1214]: would it be better to pass a shared pointer here?

File: sr_robot_lib/src/sr_motor_robot_lib.cpp [hunk 2]
```diff
@@ -1197,100 +1194,110 @@ namespace shadow_robot
...
       {
-        for (unsigned int i = 0; i < controllers_list.response.controller.size(); ++i)
+        for (auto controller : controllers_list.response.controller)
         {
-          if (ros::service::exists(controllers_list.response.controller[i].name + "/reset_gains", false))
-          {
-            ros::ServiceClient reset_gains_client = nh.template serviceClient<std_srvs::Empty>(
-              controllers_list.response.controller[i].name + "/reset_gains");
-            std_srvs::Empty empty_message;
-            if (!reset_gains_client.call(empty_message))
```
Comment (carebare47) [inline sr_robot_lib/src/sr_motor_robot_lib.cpp:1274]: Do we know what `reset_gains` did, and thus why we don't need to call it anymore?

Comment (mykolasjuraitis) [reply sr_robot_lib/src/sr_motor_robot_lib.cpp:1274]: Resetting gains reads parameters (like `max_force` and `friction_deadband`). The same parameters are being read on controller starting method so resetting should be unnecessary when starting stopped controller.

File: sr_robot_lib/src/sr_motor_robot_lib.cpp (hunk 1 above)
Comment (mykolasjuraitis) [reply sr_robot_lib/src/sr_motor_robot_lib.cpp:1214]: I am not sure it there much to gain from that. It is simple enum holder and code in this method uses thread and lamba expression so copying is the safest choice.

Comment (carebare47) [reply sr_robot_lib/src/sr_motor_robot_lib.cpp:1214]: it's a simple structure in functionality but not necessarily under the hood, if you have a look in `/home/user/projects/shadow_robot/base/devel/include/sr_robot_msgs/ControlType.h` you'll see the header file that the 
 [this .msg file](https://github.com/shadow-robot/sr_common/blob/noetic-devel/sr_robot_msgs/msg/ControlType.msg) generates

but, you know way more than I do about threads and lambda expressions, so happy to leave this at your discretion

Comment (mykolasjuraitis) [reply sr_robot_lib/src/sr_motor_robot_lib.cpp:1214]: It is also stored directly without shared pointer here: https://github.com/shadow-robot/sr-ros-interface-ethercat/blob/noetic-devel/sr_robot_lib/include/sr_robot_lib/sr_motor_robot_lib.hpp#L215

**Review by biofotis — APPROVED** [review]

File: sr_robot_lib/src/sr_motor_robot_lib.cpp [hunk 3]
```diff
@@ -65,15 +69,23 @@ namespace shadow_robot
...
                                                                  &SrMotorRobotLib::change_control_type_callback_,
                                                                  this)),
             motor_system_control_server_(
                     this->nh_tilde.advertiseService("change_motor_system_controls",
                                                     &SrMotorRobotLib::motor_system_controls_callback_,
                                                     this)),
-            lock_command_sending_(new boost::mutex())
+            joint_names_(
+            {
+              "ffj0", "ffj3", "ffj4",
```
Comment (giusebar) [inline sr_robot_lib/src/sr_motor_robot_lib.cpp]: should we account for lite hand?

**Review by giusebar — DISMISSED** [review]

File: sr_robot_lib/src/sr_motor_robot_lib.cpp (hunk 3 above)
Comment (mykolasjuraitis) [reply sr_robot_lib/src/sr_motor_robot_lib.cpp]: These are copy pasted from https://github.com/shadow-robot/sr_interface/blob/noetic-devel/sr_robot_launch/scripts/teach_mode_node#L31 so it should work as before. I assume lite hand was never supported. Happy to add more. What are joint names for lite hand?

**Pull request closed without merging**